docs(xo): add backup strategy guide (#8881)
vatesfr/xen-orchestra: docs/docs/backup_howto.md

# Backup strategy guide

This guide explains how to design and implement a backup strategy in Xen Orchestra.  
Instead of a simple list of questions and answers, it walks you through **key decisions** and **best practices** before, during, and after setting up backups.

---

## Glossary of backup types & features

- **Full backup** – Copies the entire VM each time, regardless of previous backups.  
- **Delta backup** – Saves only the changes since the last backup, reducing storage needs.  
- **Incremental replication** – Formerly called Continuous Replication (CR). Keeps a near-real-time copy of a VM on another host or pool.  
- **Full replication** – Formerly called Disaster Recovery (DR). Creates a complete replica of a VM on another host or pool.  

> - **Mirror backup** – Keeps an exact mirror of the VM backup set, with no retention—only the latest copy is kept.  

Reviewer: No mirror backup have a retention that is define on there schedule

> - **Sequence** – Groups multiple backups or replications in a specific order to be run as a single job.  

Reviewer: Not as a single job,The end of one job will trigger the following one.It's not in fact the jobs but the schedule that are running.

> - **Long-term retention** – Keeps backups over extended periods (weeks, months, or years) for compliance or archival purposes.  

Reviewer: It should be noted somewhere that this should not be mixed with schedule retention otherwise you will have backup duplication.

> - **Remote** – A storage location for backups (e.g., NFS, SMB, S3).  

Reviewer: Azure, azurite

> - **File restore** – Allows restoring individual files from a VM backup without restoring the full VM.

Reviewer: Probably should note the restrictions somewhere.

---

## 1. What should I do before setting up my backup?

Before creating your first backup job in Xen Orchestra, consider the following:

### Remote
Choose and configure your remote storage.  

> - Supported types: NFS, SMB/CIFS, S3-compatible object storage, XOSAN, or local storage.  

Reviewer: remove xosan and add azure azurite and put (not recommended ) next to local storage

- Ensure proper permissions and network connectivity.  
- Test write/read performance before relying on it.

> 

Reviewer: we should probably have something about block storage encryption and immutability here.

### Resources available
- **Storage**: Calculate how much space will be required based on your backup type and retention.  
- **Network**: Backups are network-intensive; ensure sufficient bandwidth.  
- **Compute**: Avoid backup schedules that overlap with heavy VM workloads.

### Criticality
Identify which VMs are business-critical. These should have more frequent backups and possibly replications for minimal downtime.

### Retention
Determine how long backups should be kept. This depends on:
- Compliance requirements  
- Recovery objectives (RPO/RTO)  
- Storage capacity

---

## 2. What kind of backup should I set up?

Below are the main backup and replication types available in Xen Orchestra.

### Full backup
**Advantages:**
- Complete independent copy each time  
- Simplifies restore process

**Disadvantages:**
- Requires the most storage and network bandwidth  
- Slower than incremental options

**Limitations & extensions:**
- Can be combined with long-term retention  
- No deduplication between runs

**First steps:**
1. Configure a reliable remote with sufficient capacity  
2. Schedule during low I/O periods

> 

Reviewer: Would love to have a "most suited usage section"

---

### Delta backup
**Advantages:**
- Saves only modified blocks after the first full run  
- Faster and more storage-efficient than full backups

**Disadvantages:**
- Restore requires the initial full backup plus subsequent deltas  
- Slightly more complex restore process

**Limitations & extensions:**
- Requires initial full backup  
- Can be combined with long-term retention

**First steps:**

> 1. Run an initial full backup  

Reviewer: No the first full is included in the delta backup process.

2. Schedule regular delta backups

> 

Reviewer: most suited usage section

---

### Incremental replication
**Advantages:**
- Near-continuous protection  
- Rapid failover to replicated VM

**Disadvantages:**
- Requires a secondary host or pool  
- Not suitable for long-term retention

**Limitations & extensions:**
- Works best over low-latency connections  
- Not intended as an archival method

**First steps:**

> 1. Configure target host or pool  

Reviewer: configure target storage inside the same pool or another pool

2. Ensure compatible storage and network settings

---

### Full replication
**Advantages:**

> - Creates a complete VM copy on another host/pool  

Reviewer: Creates a complete VM copy on another storage (same pool or another pool)

- Ideal for disaster recovery

**Disadvantages:**
- Longer replication times than incremental  
- Requires more bandwidth and storage

**Limitations & extensions:**
- Run less frequently than incremental replication  
- Suitable for DR plans

**First steps:**

> 1. Configure DR target host or pool  

Reviewer: Configure DR target storage (same pool or another pool)

2. Test failover procedures

---

### Sequence
**Advantages:**
- Allows chaining multiple jobs in order  
- Useful for complex backup/replication workflows

**Disadvantages:**
- Increases job complexity  
- Requires careful planning

**Limitations & extensions:**
- Limited by the capabilities of individual job types

**First steps:**
1. Identify required job order  
2. Create and test sequence on non-critical VMs

---

### Mirror backup
**Advantages:**
- Always keeps the latest backup version  

> - Saves storage by not retaining old backups

Reviewer: no mirror have retention

**Disadvantages:**

> - No historical versions available  

Reviewer: no

> - Risk of losing previous state

Reviewer: no

**Limitations & extensions:**

> - Best for non-critical or temporary workloads

Reviewer: no best suited for replication to storage with low bandwidth

**First steps:**
1. Set up a dedicated remote  
2. Test restore from latest mirror

---

## 3. What options are available to me?

### Advanced settings

> - **Smart mode** – Automatically decides whether to run a full or delta backup based on circumstances (e.g., storage conditions, snapshot issues).  

Reviewer: Absolutely not.
Smart mode is for VM selection through Tag or state in the pool

> - **Compression** – Reduces backup size at the cost of CPU usage.  

Reviewer: And backup speed

- **Encryption** – Protects backup data at rest.  

> - **Concurrency** – Controls how many jobs run in parallel.  

Reviewer: Controls how many VMs are run in // during the backup job

- **Retention policy** – Fine-tunes how many backups to keep.

---

## 4. Restore options

### VM restore
- Restore a VM to the same host/pool or another location  

> - Choose full or delta restore depending on backup type

Reviewer: not sure what this is related to but probably need to be removed

### File restore
- Access individual files within a VM backup  
- Ideal for quick recovery of deleted or corrupted files without a full VM restore

> 

Reviewer: Add a section about limitation

---

## 5. Long-term retention strategy

- Define retention periods based on compliance and operational needs  

> - Use tiered storage if possible (fast storage for recent backups, cheaper storage for older ones)  

Reviewer: no you can't define a specific storage for each type so remote is the same for all type of long term

# Backup strategy guide

This guide explains how to design and implement a backup strategy in Xen Orchestra.  

Instead of a simple list of questions and answers, it walks you through **key decisions** and **best practices** before, during, and after setting up backups.

---

## Glossary

This part explains the terminology of backup types and features.

- **Backup sequence**: A feature that allows you to chain multiple backup jobs to run one after the other, automatically. 
- [**File restore**](./backups#restore-a-file): A feature that allows you to restore individual files from a VM backup without restoring the full VM.

> - [**Full backup**](./full_backups): Copies the entire VM each time, regardless of previous backups. 

Reviewer 2: ```suggestion
- [**Full backup**](./full_backups): Copies the entire VM to  backup repositories each time, regardless of previous backups. 
```

> - [**Full replication**](./full_replication): Creates a complete replica of a VM on another host or pool.  

Reviewer 2: ```suggestion
- [**Full replication**](./full_replication): Creates a replica of a VM on other storage repositories (on the same pool or on another) by copying it completly on each run.
```

> - [**Incremental backup**](./incremental_backups): Saves only the changes since the last backup, reducing storage needs. 

Reviewer 2: ```suggestion
- [**Incremental backup**](./incremental_backups): Transfer and stores only the changes since the last backup to a backup repositories , reducing storage and network needs. The first transfer transfers the VM completly.
```

> - [**Incremental replication**](./incremental_replication): Keeps a near-real-time copy of a VM on another host or pool.

Reviewer 2: ```suggestion
- [**Incremental replication**](./incremental_replication): Transfer and stores only the changes since the last backup to a stroage repositories (on the same pool or on another), reducing network needs. The first transfer transfers the VM completly.
```

- [**Long-term retention**](./backups#long-term-backup-retention-with-gfs-strategy): Keeps backups over extended periods (weeks, months, or years) for compliance or archival purposes.  

> - [**Mirror backup**](./mirror_backup): Keeps an exact mirror of the VM backup set, with no retention. Only the latest copy is kept.  

Reviewer 2: ```suggestion
- [**Mirror backup**](./mirror_backup): mirror a backup repository to another. Retention and encryption of source and destinations can be different
```
HEre too there is differetn job to mirror full and incremental  backups

> - [**Remote**](./backups#remotes): A storage location for backups (for instance: NFS, SMB, or Amazon S3).  

Reviewer 2: ```suggestion
- [**Remote**](./backups#remotes): A storage location for backups (for instance: lcoal storage, NFS, SMB, or Amazon S3 and compatible, Microsoft Azure).  
```

---

## What should I do before setting up my backup?

Before creating your first backup job in Xen Orchestra, consider the following:

### Remote
Choose and configure your remote storage.  

> - Supported types: NFS, SMB/CIFS, S3-compatible object storage, XOSAN, or local storage.  

Reviewer 2: ```suggestion
- Supported types: NFS, SMB/CIFS, S3-compatible object storage, Microsoft Azure, or local storage.  
```
xosan is on the hosts, and I think it is not supported anymore

- Ensure proper permissions and network connectivity.  
- Test write/read performance before relying on it.

### Resources available
- **Storage**: Calculate how much space will be required based on your backup type and retention.  
- **Network**: Backups are network-intensive; ensure sufficient bandwidth.  
- **Compute**: Avoid backup schedules that overlap with heavy VM workloads.

### Criticality
Identify which VMs are business-critical. These should have more frequent backups and possibly replications for minimal downtime.

### Retention
Determine how long backups should be kept. This depends on:
- Compliance requirements
- Recovery Point Objective (RPO)
- Recovery Time Objective (RTO)  
- Storage capacity

---

## What kind of backup should I set up?

Below are the main backup and replication types available in Xen Orchestra.

### Full backup
**Pros:**
- Complete independent copy each time  

> - Simplifies restore process

Reviewer 2: ```suggestion
- Simplifies restore process
- only the allocated blocs are transfered
- can be compressed from xcp 8.2 (not sure of the exact version) 
```

**Cons:**
- Requires the most storage and network bandwidth  
- Slower than incremental options

**Limitations and extensions:**
- Can be combined with long-term retention  
- No deduplication between runs

**First steps:**
1. Configure a reliable remote with sufficient capacity  
2. Schedule during low I/O periods

---

### Delta backup
**Pros:**
- Saves only modified blocks after the first full run  
- Faster and more storage-efficient than full backups

**Cons:**
- Restore requires the initial full backup plus subsequent deltas  
- Slightly more complex restore process

**Limitations and extensions:**
- Requires initial full backup  
- Can be combined with long-term retention

**First steps:**
1. Run an initial full backup  
2. Schedule regular delta backups

---

### Incremental replication
**Pros:**

> - Near-continuous protection  

Reviewer 2: ```suggestion
- only transfer the modified data blocks since last replication 
```

- Rapid failover to replicated VM

**Cons:**
- Requires a secondary host or pool  
- Not suitable for long-term retention

**Limitations and extensions:**
- Works best over low-latency connections  

> - Not intended as an archival method

Reviewer 2: ```suggestion
- Must keep the chain short ( < 10): Not intended as an archival method
```

**First steps:**
1. Configure target host or pool  
2. Ensure compatible storage and network settings

---

### Full replication
**Pros:**
- Creates a complete VM copy on another host/pool  
- Ideal for disaster recovery

**Cons:**
- Longer replication times than incremental  
- Requires more bandwidth and storage

**Limitations and extensions:**
- Run less frequently than incremental replication  
- Suitable for DR plans

**First steps:**
1. Configure DR target host or pool  
2. Test failover procedures

---

### Pros
**Advantages:**
- Allows chaining multiple jobs in order  
- Useful for complex backup and replication workflows

**Cons:**
- Increases job complexity  
- Requires careful planning

**Limitations and extensions:**
- Limited by the capabilities of individual job types

**First steps:**
1. Identify required job order  
2. Create and test sequence on non-critical VMs

> ---

Reviewer 2: I am not sure of the context of this part

Author: The "Pros" title was wrong. This sub-section was actually about backup sequences

### Mirror backup
**Pros:**
- Always keeps the latest backup version  
- Saves storage by not retaining old backups

**Cons:**
- No historical versions available  
- Risk of losing previous state

**Limitations and extensions:**
- Best for non-critical or temporary workloads

**First steps:**
1. Set up a dedicated remote  

> 2. Test restore from latest mirror

Reviewer 2: should talk about incremental and full mirror

---

## What settings are available?

### Advanced settings
- **Smart mode**: Automatically decides whether to run a full or delta backup based on circumstances (such as storage conditions or snapshot issues)

> - **Compression**: Reduces backup size at the cost of CPU usage 

Reviewer 2: This is a setting on the backup job, and only for full

> - **Encryption**: Protects backup data at rest

Reviewer 2: this is a setting on the backup repository

> - **Concurrency**: Controls how many jobs run in parallel

Reviewer 2: ```suggestion
- **Concurrency**: Controls how many VMs run in parallel in this backup job
```

- **Retention policy**: Fine-tunes how many backups to keep over certain periods of time

---

## Restore options

### VM restore
- Restore a VM to the same host/pool or another location  
- Choose full or delta restore, depending on backup type

### File restore
- Access individual files within a VM backup  
- Ideal for quick recovery of deleted or corrupted files without a full VM restore

---

## Long-term retention strategy

- Define retention periods based on compliance and operational needs  
- Use tiered storage if possible (fast storage for recent backups, cheaper storage for older ones)  
- Regularly test restore from long-term backups

---

## Putting it all together

When designing your backup strategy with Xen Orchestra:
1. Assess criticality and resources.  
2. Choose backup types that match your RPO/RTO goals.  
3. Configure remotes and test them.  
4. Set up schedules to avoid peak loads.  
5. Apply retention policies.  
6. Test restores regularly.

---

## How to ensure XOA is always available

> Making sure XOA is always available should be a top priority for every administrator. Here’s how you can maximize its reliability:

Reviewer 2: I think the specific XO backup metadata is the xay to go to ensure you casn restart a XOA : download and install a new XOa, restore the metadata and you can restore the other backups

Since XOA runs as a virtual machine, you can apply standard VM protection measures:
- Back up regularly (full or incremental).
- Replicate the VM (full disaster recovery or incremental replication).
- Take snapshots for quick rollback if needed.

### Specific steps for the XOA VM
- Use the **XO Config** feature to back up your XOA settings. This lets you restore them to any XOA VM if necessary.

### Managing the loss of your XOA VM

If you lose the host running your XOA VM:
- **If the XOA VM was on shared storage**, you can restart it on another host in your pool.
- **If the XOA VM was stored locally** or your host was alone in its pool, deploy a new XOA VM. You can do this proactively, as there’s no limit to the number of XOA VMs in your infrastructure. Register the new VM with the same Vates account, update it, and migrate your XOA license from the old VM if needed.
- **If you are running XCP-ng 8.3**, you can use XO-lite by connecting to your master host’s IP address to manage your VMs.

### Other important considerations

> - **XOA Free** includes all VM management features without affecting your main XOA VM. All VM-level actions are handled via API calls to the pool master, so changes made through one XOA will appear across all XOAs in your infrastructure.

Reviewer 2: not sure if it's the right place
especially since backup are mostly contained in the XAO running them

(but anyt other XOA connected to the backup respositories can see the backup and restore / delete them)

- Longer replication times than incremental  
- Requires more bandwidth and storage

**Limitations and extensions:**
- Run less frequently than incremental replication  
- Suitable for DR plans

**First steps:**
1. Configure DR target storage (same pool or another pool)
2. Test failover procedures

---

### Sequence
**Advantages:**
- Allows chaining multiple jobs in order  
- Useful for complex backup and replication workflows

**Cons:**
- Increases job complexity  
- Requires careful planning

**Limitations and extensions:**
- Limited by the capabilities of individual job types

**First steps:**
1. Identify required job order  
2. Create and test sequence on non-critical VMs

---

### Mirror backup
**Pros:**
- Always keeps the latest backup version

**Limitations and extensions:**
- Best suited for replication to storage with low bandwidth
- You can mirror incremental backups or full backups

**First steps:**
1. Set up a dedicated remote  
2. Test restore from latest mirror

---

## What settings are available?

### Advanced settings
- **Smart mode**: Automatically decides whether to run a full or delta backup based on circumstances (such as storage conditions or snapshot issues)

> - **Compression**: Reduces backup size at the cost of CPU usage and backup speed.

Reviewer 2: this one is only for full backup. In this case prefer zstd if your host supports it

> - **Encryption**: Protects backup data at rest. This option is configured at the backup repository level, not per individual backup job.

Reviewer 2: this one need "sue vhd blocks" enable to work with incremental backups

- **Concurrency**: Controls how many VMs run in parallel in this backup job
- **Retention policy**: Fine-tunes how many backups to keep over certain periods of time

---

## Restore options

### VM restore
- Restore a VM to the same host/pool or another location  

> - Choose full or delta restore, depending on backup type

Reviewer 2: differential restore is only available for incremental backup , it's on a best effort basis, since there are  lot of internal prerequisite to use it on the backups and snapshot chains, storage 
If it's not possible, it automatically fall back to a normal restore , transferring the full VM to the hosrs

For full backup, it will always be a full restore

### File restore
- Access individual files within a VM backup  
- Ideal for quick recovery of deleted or corrupted files without a full VM restore

### Limitations

> ---

Reviewer 2: only the simpler partition types are supported: those that can mounted on a debian + NTFS. Especially ReFS and dynamic disks on windows are completly unsupported.  LVM support is partial, and have a lot of edge case for now. 

It is intended to be used to restore a small set of file

for the more complex cases, we provide a script that allow a user to mount the backup chain as a raw disk, and then do the partition discovery and mouting by hand https://github.com/vatesfr/xen-orchestra/tree/master/%40vates/fuse-vhd

## Long-term retention strategy

- Define retention periods based on compliance and operational needs  
- Regularly test restore from long-term backups

:::warning

> To prevent backup duplication, do not mix long-term retention with schedule retention.

Reviewer 2: ```suggestion
To prevent backup duplication, do not mix long-term retention with multiple schedules.
```

---

### Mirror backup
**Pros:**
- Always keeps the latest backup version

**Limitations and extensions:**
- Best suited for replication to storage with low bandwidth
- You can mirror incremental backups or full backups

**First steps:**
1. Set up a dedicated remote  
2. Test restore from latest mirror

---

## What settings are available?

### Advanced settings

- **Compression**: Reduces backup size at the cost of CPU usage and backup speed.
- **Encryption**: Protects backup data at rest.
- **Concurrency**: Controls how many VMs run in parallel in this backup job.
- **Retention policy**: Fine-tunes how many backups to keep over certain periods of time.

:::tip
- **Compression** is configured at the backup job level, and applies only to full backups. For full backups, prefer [Zstandard (Zstd)](https://en.wikipedia.org/wiki/Zstd) if your host supports it.
- **Encryption** is configured at the backup repository level, not per individual backup job. To use encryption with incremental backups, the **use VHD blocks** setting must be enabled.
:::

---

## Restore options

### VM restore
- Restore a VM to the same host/pool or another location.

> - When restoring, Xen Orchestra can attempt a **delta restore**, which reuses the current VM disk to speed up the process.

Reviewer 2: ```suggestion
- When restoring, Xen Orchestra can attempt a **differential restore**, which reuses the current VM disk to speed up the process.
```
<img width="610" height="415" alt="image" src="https://github.com/user-attachments/assets/2cb843f2-292a-4ce0-a823-e21499a73277" />
